### DISCUSSION TITLE
Admin list filters have become hard due to parser integration

### DISCUSSION THREAD

**Discussion by giannik** [OrchardCMS/OrchardCore#10572] [Q&A]
Im trying to build filters in the admin based on a custom index table call Products and Orders for an e-commerce type application.
In the Products index table there is a price field (numeric field) and and HasDiscount field (boolean) that are mapped to orchard content fields (Numeric field and Boolean field).
The  tables are populated just fine. 
Admin users must be able  from the admin to filter products based on price (eg  Price > 100 etc)
Now in the admin its very hard to build admin filters for the price fields. 
My main obsticle is the parser integration recently added. 
I tried adding a NumericParser  similar to the DateTimeParser added in the the audittrail  module by @deanmarcussen .
But it failed , as I dont know how parser logic works. 
Its kinda frusturating to have to  build a parser to add a filter in the admin based on  a field in a custom index table.
Also frusturating is that some other admin filters (like taxonomy admin filters) dont us the parser logic.
So its not clear for developers the direction to follow. 
On one hand i think filtering content in the admin is very important based on your custom index tables especially for more business oriented apps like ecommerce apps 
On the other , a developer should have a clear direction on how to do this filtering from the admin filters feature.
**Writing your own parser should not be a must.**
So I think a step in the right direction would be to : 

- Abstract the parser logic into  some core module at least for the most common field types (like Datetime as @deanmarcussen  has done as well as Numeric Fields , Boolean fields and maybe string fields. )
- Maybe also adapt other admin filters (like the localization admin filters and taxonomy admin filters ) to use the some filtering logic. 

I would be willing to sponsor such a feature as I think its a showstopper for building more sophisticated apps with orchard.

**Comment by Skrypt** [comment]
You can do this with Lucene Queries. Maybe it's not what you want, because it's not filtering data from SQL Server directly from a complex SQL Query. But in the end, the Lucene Query will be faster.

**Comment by giannik** [comment]
No, I dont want to use lucene .
Admin filters are great for developers to build and end users to consume.
Its just that the parser integration is not fully adapted and not many developers know how to build parsers.
Its very easy to build your own UI with DisplayDriver < ContentOptionsViewModel >  and populate the UI with your own custom filters and implement a custom IContentsAdminListFilterProvider.
And filtering for numbers works from the admin but only for equality operators.  You need to write a numeric parser to handle other operators.
@deanmarcussen did this for dates in the audit trail so  it would be great to re-use this on content items filters.

**Comment by deanmarcussen** [comment]
I would actually suggest that lucene is quite a good solution for this, as the parser / filter arrangement was built with enough fleixibility that you could send the search request of to lucene, and use it.

However you don't want to use Lucene, so you need a filter parser.
Noting that before there was a filter parser there would be no way to do what you are wanting to do.

The old filter arrangement still exists for taxonomies, as I have not had time to work on it, but the old filter arrangement would never have supported `> 10`.

The only way to support something like `> 0` (which then has all the other combinations of < 0, >= etc al is to have something that parses these and converts them into a query.

I'm quite happy to assist in writing a number parser.

Suggest you could help by defining the syntax that you would want to support.

**Comment by giannik** [comment]
thanks @deanmarcussen 

here is a repo on what Im trying to acheive.

https://github.com/giannik/OrchardCore/tree/numeric_parser

If you run this and go to the admin there are 2  content items  of contenttype = resume (just add in the contenttype picker the resume content type).
You will see the 2 custom admin filters that map to a custom index table .
If you type in the Rate Per Hour filter an absosulte value  (like 100 or 200) it filters correctly as shown below :


![image](https://user-images.githubusercontent.com/1516981/139288691-4741e110-754b-4fe2-aeb7-39a180104c51.png)


But I would like to support in the text box to write :  

>100
>=100
<100
<=100
not 100
100..250  to map range values.

So this is where a parser is needed. 


At this line :
https://github.com/giannik/OrchardCore/blob/1a082709d0a81b8979d6fad1e35620e2781704bd/src/OrchardCore.Modules/Intelli.Resumes/Services/ResumeContentsAdminListFilterProvider.cs?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L29

is my attempt to write a numeric parser but it does not work , because  I dont know anything how a parser works.


Oh the credentials to login the admin are : 
admin
Admin123!@#
